### PR TITLE
Add placeholder to Post Comments block

### PR DESCRIPTION
**Pull request by luisherranz**
## What?

This PR adds a placeholder in the Editor for the Post Comments block. It includes:

- A warning at the top with the text:

  > This is just a placeholder, not a real comment. The final styling may differ because it also depends on the current theme. For better compatibility with the Block Editor, please consider replacing this block with the "Comments Query Loop" block.

- The navigation links
- A comment example, including an avatar, author name, date, content, and reply link.
- The reply form.

## Why?

Before this, the Post Comments blocked rendered:

- Site Editor: A warning with the text `Post comments block: no post found.`
- Post Editor:
  - Post with comments: A loop of the raw HTML of all the comments, unnested and unstyled.
  - Post without comments: The text `No comments`.

None of those are especially useful, considering the only action users can perform in this block is to configure its style.

## How?

I've added the default comments markup. The only dynamic parts are the post title and the user name (to simulate the reply form message).

I've used Gutenberg's `<Warning>` component for the warning. I'm not a big fan of how it looks like in the UI, but I think it's out of the scope of this PR to address that.

I acknowledge that this type of placeholder is far from ideal, but @darerodz and I analyzed the other possibilities, and we concluded that it was the least bad option. The other possibilities we considered were:

- Using `<ServerSideRender>` component.

  Executing `comment_template()` didn't work within a REST API call (`/wp/v2/block-renderer/`) because it doesn't have a post context.

  Even if that could be solved, the result wouldn't faithfully represent the frontend because:

  - The comment-related styles from the theme and plugins would be missing.
  - It would not work in the site editor because we can't inject dummy data to the `comment_template()` function.

- Looping over the real comments in the Editor to reproduce the final HTML as faithfully as possible.

  Replicating the logic of the comments is not trivial (as [we've seen in the Comments Query Loop blocks](https://github.com/WordPress/gutenberg/issues/34994)).

  Also, the result wouldn't faithfully represent the frontend because:

  - The markup modifications of the theme and plugins would be missing.
  - The comment-related styles from the theme and plugins would be missing.

So at the end, and considering that this is a temporary block that should be removed once the [Comments Query Loop](https://github.com/WordPress/gutenberg/issues/34994) and the [Comments Form blocks](https://github.com/WordPress/gutenberg/issues/38107) are mature enough, we decided to add the default comments markup so the user can get a notion of the styles, and a warning to prevent confusion.

## Testing Instructions

Site Editor:

1. Add the Post Comments block to a template.
2. Look that the placeholder looks as expected.
3. Test that the styling options work.

Post Editor:

1. Add the Post Comments block to a post.
2. Look that the placeholder looks as expected.
3. Test that the styling options work.

## Screenshots or screencast 

<a href="https://www.loom.com/share/91b518264de84e75a7c3a64e283ee0c2">
    
  </a>

https://www.loom.com/share/91b518264de84e75a7c3a64e283ee0c2

### CODE REVIEW

File: packages/block-library/src/post-comments/edit.js
Comment (michalczaplinski) [inline packages/block-library/src/post-comments/edit.js]: Is it okay if we depend on a link to gravatar? 🤔

File: packages/block-library/src/post-comments/edit.js
Comment (michalczaplinski) [inline packages/block-library/src/post-comments/edit.js]: Not sure if we should remove this part here or if I should update https://github.com/WordPress/gutenberg/pull/40368 to include the "Logged in as {currentUser}". I was following https://github.com/WordPress/gutenberg/issues/26999#issuecomment-1099585096 in the overall design because if the user is not logged in the "Logged in as {currentUser}" line is not going to be displayed on the frontened.

**Review by michalczaplinski — COMMENTED** [review]
Nice job, I think it looks great already! 👍 I only have some minor comments here

File: packages/block-library/src/post-comments/edit.js
Comment (luisherranz) [reply packages/block-library/src/post-comments/edit.js]: It should be theoretically safe because it's what WordPress is using on new installations, but I agree it'd be better to have something that is not hardcoded.

I'll use the default avatar exposed in the state 🙂

File: packages/block-library/src/post-comments/edit.js
Comment (luisherranz) [reply packages/block-library/src/post-comments/edit.js]: Sure. I'll remove it (the simplification proposed by @paaljoachim in https://github.com/WordPress/gutenberg/issues/26999#issuecomment-1099585096) to keep the consistency between the placeholders.

image

File: packages/block-library/src/post-comments/edit.js
Comment (luisherranz) [reply packages/block-library/src/post-comments/edit.js]: Done in https://github.com/WordPress/gutenberg/pull/40484/commits/b3b2b6c824d6cb153ee3b0b17e0f7d9a05ee5d64.

File: packages/block-library/src/post-comments/edit.js
Comment (luisherranz) [reply packages/block-library/src/post-comments/edit.js]: Done in https://github.com/WordPress/gutenberg/pull/40484/commits/fe62c6569b367ed2382772c863635362cc11114a.

**Review by SantosGuillamot — APPROVED** [review]
It looks great to me! 🙂

Would it make sense to add some css to prevent interactivity in the editor? I mean, as the links don't work and the `textarea`, for example, doesn't accept input, maybe we could prevent the clicks there with something like:

```
.wp-block-post-comments * {
    pointer-events: none;
}
```

**Comment by luisherranz** [comment]
Thanks, Mario. I'll add that CSS 🙂

**Comment by cbravobernal** [comment]
Maybe adding `disabled` to the input would be a better option. I'm not sure if with pointer-events: none, iOS Voiceover seems that cannot access the inner parts of the elements of the form to see what you are going to render.

**Comment by DAreRodz** [comment]
For me, this looks great so far. Thaks for taking care of this, @luisherranz! 🙌

PS: it's a bit weird to see the warning message styles changing when you modify, e.g., the comments color or font size. 😅 Not sure if we could/should do something regarding that.

**Comment by luisherranz** [comment]
> PS: it's a bit weird to see the warning message styles changing when you modify, e.g., the comments color or font size. 😅 Not sure if we could/should do something regarding that.

Yes, that should be definitely fixed, but probably out of the scope of this PR.

EDIT: I've opened a separate issue for that https://github.com/WordPress/gutenberg/issues/40536.

> Maybe adding disabled to the input would be a better option.

~~I already added `readOnly` because I saw @michalczaplinski did it here https://github.com/WordPress/gutenberg/pull/40368/commits/d7650b20dba549d69b2d5e7de0e20a934cfd2feb. If `disabled` is also recommended let me know.~~

EDIT: Removed in favor of `useDisabled` as described in https://github.com/WordPress/gutenberg/pull/40484#issuecomment-1105393780.

---

I saw in @michalczaplinski's PR (https://github.com/WordPress/gutenberg/pull/40368) a bit of logic to show different warning messages if the post you're editing doesn't support comments for some reason, so I've added it here as well. It took it a bit further, though.

@michalczaplinski could you please check it out, let me know what you think, and maybe if you like the approach, merge this and reuse the same logic for your block (create a hook that can be shared, or something like that).

This is the flow I've used:



_[Excalidraw](https://excalidraw.com/#json=M0gu0PwQkviXEZkfzp-N0,ga2JQ4hT1tp9aMhlRV1-tQ)_

And this is the video explanation:

<a href="https://www.loom.com/share/339af4d5ee974758a3c8f6b491f2bf0d">
    
  </a>

https://www.loom.com/share/339af4d5ee974758a3c8f6b491f2bf0d

**Comment by ockham** [comment]
Thanks a lot @luisherranz, this is looking great!

@priethor has [asked](https://github.com/WordPress/gutenberg/pull/40368#pullrequestreview-948498814) over at #40368 for some design feedback on how to make it even clearer that these placeholders aren't really editable, to which @jasmussen kindly [replied](https://github.com/WordPress/gutenberg/pull/40368#issuecomment-1105116574).

Would it be possible to apply Joen's suggestions to this PR as well? 😊 🙏

**Comment by luisherranz** [comment]
> Would it be possible to apply Joen's suggestions to this PR as well?

I'm on it 🙂

**Comment by luisherranz** [comment]
I've implemented it as @darerodz described in https://github.com/WordPress/gutenberg/pull/40368#issuecomment-1105226653, and it seems to work great. 

I've also added @santosguillamot's suggestion https://github.com/WordPress/gutenberg/pull/40484#pullrequestreview-948166149. 

Now the placeholder looks almost like an image and can't be selected, but you can still see the styles changes.

https://user-images.githubusercontent.com/3305402/164498088-7f307f0f-d3c3-4a95-aa42-befcb29c137c.mp4

**Review by cbravobernal — APPROVED** [review]
LGTM! But still have some doubts if how accessible is this block for screen readers.

**Review by ockham — APPROVED** [review]
Thanks @luisherranz! This is looking great 👍 

Let's merge this as-is.

Some ideas for follow-ups:

- We might be able to figure out the CSS quirks you mentioned in your Loom in a follow-up.
- I also don't know how this is going to work in terms of a11y. Maybe we can have someone who has a better grasp of a11y review this, so we can fix it in a follow-up.
- We'll go with the Warning for now (even though this is likely not an established UX pattern for this). However, with #40521, the placeholder will only be used for an empty Post Comments block, so we'll change the warning there to reflect that.
- Maybe we can re-use the Comment Form block placeholder (#40368) to render the form here (and de-dupe code), once that's merged.
- We could consider making the placeholder resemble the Comments Query Loop a bit more -- specifically e.g. by using the post's actual comments when inside the Post Editor.

**Comment by ockham** [comment]
👋  @ryelle If you can spare some time, could we have your a11y feedback on this PR? It basically adds a placeholder for a post's comments (and comment form) to the editor which cannot really be edited -- really just to have something that somewhat captures what it'll look like on the frontend. We were a bit unsure if this will surface to screenreaders in an undesired way 😅 

You need this patch to re-enable the (currently deprecated) Post Comments block:

```diff
diff --git a/packages/block-library/src/post-comments/block.json b/packages/block-library/src/post-comments/block.json
index 6c5c4e16b8..8a15e57569 100644
--- a/packages/block-library/src/post-comments/block.json
+++ b/packages/block-library/src/post-comments/block.json
@@ -33,8 +33,7 @@
                                "background": true,
                                "text": true
                        }
-               },
-               "inserter": false
+               }
        },
        "style": [
                "wp-block-post-comments",
```

Then, insert the Post Comments block in the Post or Site editor to see the placeholder.

**Comment by ockham** [comment]
(Merging as discussed [above](https://github.com/WordPress/gutenberg/pull/40484#pullrequestreview-948942959); additional feedback to be addressed in follow-up PRs.)

**Comment by ryelle** [comment]
@ockham thanks for the ping 😄  I did some quick testing - I hope this helps.

When I turned on VoiceOver and looked at the different ways to navigate through the page, the content in the Post Comment block shows up. Like in the heading list, the “One response…” and “Leave a Reply” are there, when other headings in the page aren't (because heading blocks have different roles). I think this might be a larger gutenberg issue though, and I don't know what the expected behavior should be.

When the block is added, the first thing I hear is the warning, and then it’s possible to navigate through the whole comment section. I think this is good, since the user would want to know what’s in there. When navigating by keyboard or VoiceOver + keyboard, it’s possible to focus on the links, and activating a link reloads the page (pretty sure the root issue is this https://github.com/WordPress/gutenberg/issues/15799). The comment textarea and button are disabled, as expected.

The textarea has no label, not sure if that’s from the theme or something to fix in the block.

I was able to change the styles with keyboard & VoiceOver, but tweaking the typography settings (like letter spacing) also changed the warning. Not an a11y issue, just unexpected 🙂 

Screen Shot 2022-04-21 at 1 57 37 PM

**Comment by luisherranz** [comment]
Thanks a lot for the review, @ryelle!

> The textarea has no label, not sure if that's from the theme or something to fix in the block.

That's actually my fault. The label is there, but I removed the textarea id because the linter was complaining. I'll fix it in a new PR. Thanks for the heads up!

EDIT: PR created https://github.com/WordPress/gutenberg/pull/40527

> VoiceOver + keyboard, it's possible to focus on the links, and activating a link reloads the page (pretty sure the root issue is this #15799) 

It seems like it, yes.

> tweaking the typography settings (like letter spacing) also changed the warning

Yes, @darerodz also mentioned it here https://github.com/WordPress/gutenberg/pull/40484#issuecomment-1104884198. I haven't found any issue that describes that problem, so I'll open one.

EDIT: Issue created https://github.com/WordPress/gutenberg/issues/40536.

> heading list, the “One response…” and “Leave a Reply” are there, when other headings in the page aren't (because heading blocks have different roles). I think this might be a larger gutenberg issue though, and I don't know what the expected behavior should be.

Do you know about any other block that contains similar headers, but they don't appear? (to take a look at the code and see if they are doing something different).

**Comment by ockham** [comment]
Thanks a lot for your a11y feedback, @ryelle -- much appreciated!
Thanks @luisherranz for addressing it!

File: packages/block-library/src/post-comments/edit.js
Comment (adamziel) [inline packages/block-library/src/post-comments/edit.js:102]: Nit: the "e" is missing from showPlac**e**holder

**Comment by gziolo** [comment]
Cherry-picked to the `wp/6.0` branch for WordPress Beta 3 release.

File: packages/block-library/src/post-comments/edit.js
Comment (luisherranz) [reply packages/block-library/src/post-comments/edit.js:102]: Thanks, Adam. ~~I'll fix it in a new PR~~.

**Comment by luisherranz** [comment]
Backport of the exposure of the `default_comments_status` option: https://github.com/WordPress/wordpress-develop/pull/2623

File: packages/block-library/src/post-comments/edit.js
Comment (luisherranz) [reply packages/block-library/src/post-comments/edit.js:102]: It was also wrong in the [Post Comments Form PR](https://github.com/WordPress/gutenberg/pull/40563#pullrequestreview-952014022), so I've changed it there for both blocks: https://github.com/WordPress/gutenberg/pull/40563/commits/b575b9454c1466547c2b23783dd820ee4e264f21

**Pull request merged**
